File: alg_ds_1/ordered_list/ordered_list.go
package main

import (
	"constraints"
	"errors"
	"os"
	"strings"
)

type Node[T constraints.Ordered] struct {
	prev  *Node[T]
	next  *Node[T]
	value T
}

type OrderedList[T constraints.Ordered] struct {
	head       *Node[T]
	tail       *Node[T]
	_ascending bool
	count      int
}

func (l *OrderedList[T]) Count() int {
	return l.count
}

func (l *OrderedList[T]) Add(item T) {
	l.count++
	newNode := Node[T]{value: item}
	if l.head == nil {
		l.head = &newNode
		l.tail = &newNode
		return
	}
	n := l.head
	for n != nil {
		if (l._ascending && l.Compare(n.value, item) >= 0) || (!l._ascending && l.Compare(n.value, item) <= 0) {
			l.addToMiddle(n, &newNode)
			return
		}
		n = n.next
	}
	newNode.prev = l.tail
	l.tail.next = &newNode
	l.tail = &newNode
}

func (l *OrderedList[T]) addToMiddle(n, newNode *Node[T]) {
	if n == l.head {
		newNode.next = l.head
		l.head.prev = newNode
		l.head = newNode
	} else {
		newNode.prev = n.prev
		n.prev.next = newNode
		n.prev = newNode

		newNode.next = n
	}
}

func (l *OrderedList[T]) Find(n T) (Node[T], error) {
	node := l.head
	for node != nil {
		if (l._ascending && l.Compare(node.value, n) > 0) || (!l._ascending && l.Compare(node.value, n) < 0) {
			return Node[T]{value: n, next: nil, prev: nil}, errors.New("node not found")
		}
		if l.Compare(node.value, n) == 0 {
			return *node, nil
		}
		node = node.next
	}
	return Node[T]{value: n, next: nil, prev: nil}, errors.New("node not found")
}

func (l *OrderedList[T]) Delete(n T) {
	node := l.head

	for node != nil {
		if l.Compare(node.value, n) != 0 {
			node = node.next
			continue
		}

		if node == l.head {
			l.head = l.head.next
		}

		if node.prev != nil {
			node.prev.next = node.next
		}

		if node.next != nil {
			node.next.prev = node.prev
		} else {
			l.tail = node.prev
		}

		if l.head == nil {
			l.tail = nil
		}

		l.count--
		return
	}
}

func (l *OrderedList[T]) Clear(asc bool) {
	l._ascending = asc
	l.head, l.tail = nil, nil
	l.count = 0
}

func (l *OrderedList[T]) Compare(v1 T, v2 T) int {
	val1 := interface{}(v1)
	val2 := interface{}(v2)

	s1, ok1 := val1.(string)
	s2, ok2 := val2.(string)

	if ok1 && ok2 {
		return l.CompareStrings(s1, s2)
	}

	if v1 < v2 {
		return -1
	}
	if v1 > v2 {
		return +1
	}
	return 0
}

func (l *OrderedList[T]) CompareStrings(s1, s2 string) int {
	s1 = strings.TrimSpace(s1)
	s2 = strings.TrimSpace(s2)
	if s1 < s2 {
		return -1
	}
	if s1 > s2 {
		return 1
	}
	return 0
}

File: alg_ds_1/python_version/07_ordered_list/ordered_list.py
class Node:

    def __init__(self, value):
        self.value = value
        self.prev = None
        self.next = None


class OrderedList:

    # Задание на курсе: 7
    # Задача: 1
    # Название: создание упорядоченного списка
    def __init__(self, asc):
        self.head = None
        self.tail = None
        self.__ascending = asc

    # Задание на курсе: 7
    # Задача: 2
    # Название: сравнение числовых значений
    # Временная сложность: O(1)
    # Пространственная сложность: O(1)
    def compare(self, v1, v2):
        if v1 < v2:
            return -1

        if v1 > v2:
            return 1

        return 0

    # Задание на курсе: 7
    # Задача: 3
    # Название: добавление значения с сохранением порядка
    # Временная сложность: O(n)
    # Пространственная сложность: O(1)
    def add(self, value):
        new_node = Node(value)
        current = self.head

        while current is not None:
            comparison = self.compare(current.value, value)
            should_insert_before = (
                self.__ascending and comparison >= 0
                or not self.__ascending and comparison <= 0
            )

            if should_insert_before:
                self._insert_before(current, new_node)
                return

            current = current.next

        self._append_node(new_node)

    # Задание на курсе: 7
    # Задача: 4
    # Название: удаление первого элемента по значению
    # Временная сложность: O(n)
    # Пространственная сложность: O(1)
    def delete(self, val):
        node = self.find(val)

        if node is None:
            return

        self._unlink_node(node)

    # Задание на курсе: 7
    # Задача: 6
    # Название: поиск элемента с ранним прерыванием
    # Временная сложность: O(n)
    # Пространственная сложность: O(1)
    def find(self, val):
        current = self.head

        while current is not None:
            comparison = self.compare(current.value, val)

            if comparison == 0:
                return current

            passed_value = (
                self.__ascending and comparison > 0
                or not self.__ascending and comparison < 0
            )

            if passed_value:
                return None

            current = current.next

        return None

    # Название: очистка списка и изменение направления сортировки
    # Временная сложность: O(1)
    # Пространственная сложность: O(1)
    def clean(self, asc):
        self.head = None
        self.tail = None
        self.__ascending = asc

    # Название: количество элементов в списке
    # Временная сложность: O(n)
    # Пространственная сложность: O(1)
    def len(self):
        count = 0
        current = self.head

        while current is not None:
            count += 1
            current = current.next

        return count

    # Название: получение всех узлов списка
    # Временная сложность: O(n)
    # Пространственная сложность: O(n)
    def get_all(self):
        nodes = []
        current = self.head

        while current is not None:
            nodes.append(current)
            current = current.next

        return nodes

    def _is_ascending(self):
        return self.__ascending

    def _append_node(self, node):
        node.prev = self.tail
        node.next = None

        if self.tail is None:
            self.head = node
            self.tail = node
            return

        self.tail.next = node
        self.tail = node

    def _insert_before(self, current, new_node):
        previous = current.prev
        new_node.prev = previous
        new_node.next = current
        current.prev = new_node

        if previous is None:
            self.head = new_node
            return

        previous.next = new_node

    def _unlink_node(self, node):
        previous = node.prev
        following = node.next

        if previous is None:
            self.head = following

        if previous is not None:
            previous.next = following

        if following is None:
            self.tail = previous

        if following is not None:
            following.prev = previous

        node.prev = None
        node.next = None


# Задание на курсе: 7
# Задача: 5
# Название: упорядоченный список строк
class OrderedStringList(OrderedList):

    # Временная сложность: O(k)
    # Пространственная сложность: O(k)
    def compare(self, v1, v2):
        return super().compare(v1.strip(), v2.strip())


